collector/lib/ResourceSelector.cpp

```diff
@@ -0,0 +1,138 @@
+//
+// Created by Robby Cochran on 3/28/24.
+//
+
+#include "ResourceSelector.h"
+
+#include <regex>
+
+#include "Hash.h"
+
+namespace collector {
+
+UnorderedMap<std::string, std::regex> ResourceSelector::regexMap_;
+
+std::regex ResourceSelector::GetOrCompileRegex(const std::string& str) {
+  auto pair = regexMap_.find(str);
+  if (pair != regexMap_.end()) {
+    return pair->second;
+  } else {
+    std::regex pattern(str);
+    regexMap_.insert(std::make_pair(str, pattern));
+    return pattern;
+  }
+}
+
+bool ResourceSelector::IsRuleValueFollowed(const storage::RuleValue& value, const std::string& resource_name) {
+  if (value.match_type() == storage::REGEX) {
+    std::regex pattern = GetOrCompileRegex(value.value());
+    return std::regex_match(resource_name, pattern);
+  } else {
+    return (resource_name == value.value());
+  }
+}
+
+bool ResourceSelector::IsRuleFollowed(const storage::SelectorRule& rule, const std::string& resource_name) {
+  const auto& values = rule.values();
+
+  if (values.size() == 0) {
+    return true;
+  }
+
+  auto op = [resource_name](const auto& value) -> bool {
+    return IsRuleValueFollowed(value, resource_name);
+  };
+
+  if (rule.operator_() == storage::BooleanOperator::OR) {
+    return std::any_of(values.begin(), values.end(), op);
+  } else if (rule.operator_() == storage::BooleanOperator::AND) {
+    return std::all_of(values.begin(), values.end(), op);
+  }
+
+  return false;
+}
+
+bool ResourceSelector::IsResourceInResourceSelector(const storage::ResourceSelector& rs, const std::string& resource_type, const std::string& resource_name) {
+  // const auto& rules = rs.rules();
+
+  // auto op = [resource_type, resource_name](const auto& rule) -> bool {
+  //   if (rule.field_name() == resource_type) {
+  //     if (!IsRuleFollowed(rule, resource_name)) {
+  //       return false;
+  //     }
+  //     return true;
+  //   }
+  //   return true;
+  // };
+
+  // return std::all_of(rules.begin(), rules.end(), op);
+
+  for (const auto& rule : rs.rules()) {
+    if (rule.field_name() == resource_type) {
+      if (!IsRuleFollowed(rule, resource_name)) {
+        return false;
+      }
+    }
+  }
+
+  return true;
```

Comment:
I see that you tried to turn this one to use `std::all_of` but ultimately went back on it, why is that? I think it would be ok to use:
```cpp
return std::all_of(rules.begin(), rules.end(), [resource_type, resource_name](const auto& rule) -> bool {
  return rule.field_type() != resource_name || IsRuleFollowed(rule, resource_name);
});
```

Reply:
I don't know what I was thinking with this one. Your suggestion here is good and I am going with it.